File: iCyption2022/CftTest/holy/solution.py
ROLES_COUNT = 6
roles = ["Saber", "Lancer", "Archer", "Caster", "Assassin", "Berserker"]

dummyFormation = { "fmt": {}, "value" : 0, "cost" : 0, "cnt": 0}

for role in roles :
    dummyFormation["fmt"][role] = None

heroes = {"Saber" : [], "Lancer" : [], "Archer" : [], "Caster" : [], "Assassin" : [], "Berserker" : []}

n = int(input())

for i in range(n) :
    name, role, skillValue, price = map(str, input().split())
    skillValue, price = int(skillValue), int(price)
    heroes[role].append({"name": name, "role": role, "skillValue": skillValue, "price": price})

budget = int(input())

dpFormation = [dummyFormation for i in range(budget + 1)]

for i in range(ROLES_COUNT) :
    prevDPFormation = dpFormation
    
    for j in range(len(heroes[roles[i]])) :
        currBudget = budget
        currHero = heroes[roles[i]][j]

        while (currBudget >= currHero["price"]) :
            prevFormation = prevDPFormation[currBudget - (currHero["price"])]
            if (prevFormation["cnt"] == i) :
                newFormation = prevFormation
                
                newFormation["fmt"][currHero["role"]] = heroes[roles[i]][j]
                newFormation["value"] += currHero["skillValue"]
                newFormation["cost"] += currHero["price"]
                newFormation["cnt"] += 1

                if ((dpFormation[currBudget]["cnt"] <= i) or (dpFormation[currBudget]["value"] <= newFormation["value"])) :
                    dpFormation[currBudget] = newFormation

            currBudget -= 1

bestFormation = dpFormation[budget]
valid = (True if bestFormation["cnt"] == ROLES_COUNT else False)

if (valid) :
    print(bestFormation["value"])
else :
    print("Tidak bisa ikut :(")

File: iCyption2022/CftTest/holy/solution.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

struct hero {
    string name;
    string role;
    int skillValue;
    int price;
};

typedef struct hero *ptrHero;

struct formation {
    map<string, ptrHero> fmt;
    int value = 0;
    int cost = 0;
    int cnt = 0;
};

const int ROLES_COUNT = 6;
const string roles[ROLES_COUNT] = { "Saber", "Lancer", "Archer", "Caster", "Assassin", "Berserker" };

int main() {
    int n;
    int budget, skillValue, price;

    formation dummyFormation;
    for (string role : roles) dummyFormation.fmt[role] = nullptr;

    string name, role;

    map<string, vector<hero>> heroes;

    cin >> n;

    for (int i = 0; i < n; i++) {
        cin >> name >> role >> skillValue >> price;
        heroes[role].push_back({name, role, skillValue, price});
    }
    
    cin >> budget;

    vector<formation> dpFormations(budget + 1, dummyFormation);

    for (int i = 0; i < ROLES_COUNT; i++) {
        vector<formation> prevDPFormations(dpFormations);

        for (int j = 0; j < (int)heroes[roles[i]].size(); j++) {
            int currBudget = budget;
            hero currHero = heroes[roles[i]][j];

            while (currBudget >= currHero.price) {
                formation prevFormation = prevDPFormations[currBudget - (currHero.price)];
                if (prevFormation.cnt == i) {
                    formation newFormation = prevFormation;

                    newFormation.fmt[currHero.role] = &heroes[roles[i]][j];
                    newFormation.value += currHero.skillValue;
                    newFormation.cost += (currHero.price);
                    newFormation.cnt++;

                    if ((dpFormations[currBudget].cnt <= i) || (dpFormations[currBudget].value <= newFormation.value)) {
                        dpFormations[currBudget] = newFormation;
                    }
                }

                currBudget--;
            }
        }
    }

    formation bestFormation;

    bestFormation = dpFormations[budget];
    bool valid = bestFormation.cnt == ROLES_COUNT;

    if (valid) {
        cout << bestFormation.value << endl;
    } else {
        cout << "Tidak bisa ikut :(" << endl;
    }
}